sympy/polys/matrices/ddm.py

```diff
@@ -92,6 +92,21 @@ def __init__(self, rowslist, shape, domain):
         if not (len(self) == m and all(len(row) == n for row in self)):
             raise DDMBadInputError("Inconsistent row-list/shape")
 
+    def extract_slice(self, slice1, slice2):
+        m, n = self.shape
+        ri = range(m)[slice1]
+        ci = range(n)[slice2]
+
+        ddm = []
+        for i, row in enumerate(self):
+            if i in ri:
+                ddm.append([e for j, e in enumerate(row) if j in ci])
```

Comment:
Just use the slice1 and slice2 directly.

Reply:
I showed above a different way of implementing slicing for `DDM`: https://github.com/sympy/sympy/pull/21118#discussion_r597610726

Have you done timings to compare the different approaches?

Reply:
Most probably, your approach will be faster as it involves single for loop and slicing.

But, there are two drawbacks
It doesn't work for negatives slices like `dm.extract_slice(slice(1, 2), slice(-3))`
It won't handle returning a single item as DomainScalar.  

The below implementation seems to work resolve the 2 issues and uses the faster implementation as suggested:
```
def extract_slice(self, slice1, slice2):
        from .domainmatrix import getindex, DomainScalar
        m, n = self.shape
        i, _m = getindex(slice1, m)
        j, _n = getindex(slice2, n)

        if i >= m:
            raise IndexError("row index out of range")
        if j >= n:
            raise IndexError("column index out of range")

        slice1, slice2 = slice(i, _m), slice(j, _n)
        ddm = [row[slice2] for row in self[slice1]]

        if (len(ddm), len(ddm[0])) == (1, 1):
            return DomainScalar(ddm[0][0], self.domain)

        return DDM(ddm, (len(ddm), len(ddm[0])), self.domain)
```

Some passed test cases:
```
def test_DDM_extract_slice():
    dm = DDM([
        [ZZ(1), ZZ(2), ZZ(3)],
        [ZZ(4), ZZ(5), ZZ(6)],
        [ZZ(7), ZZ(8), ZZ(9)]], (3, 3), ZZ)

    assert dm.extract_slice(slice(0, 3), slice(0, 3)) == dm
    assert dm.extract_slice(slice(1, 2), slice(-3)) == DDM([[4], [7]], (2, 1), ZZ)
    assert dm.extract_slice(slice(2, 2), slice(-3)) == DomainScalar(ZZ(7), ZZ)
    assert dm.extract_slice(slice(-2), slice(-3)) == DDM([[1], [4]], (2, 1), ZZ)
    assert dm.extract_slice(slice(-2), slice(0, 1)) == DDM([[1, 2], [4, 5]], (2, 2), ZZ)
```

@oscarbenjamin, Kindly share your opinion.
Thanks.

Reply:
> Most probably, your approach will be faster

Rather than guessing can you do some timings?

Reply:
> Rather than guessing can you do some timings?
Sorry about that:

Timings for the code already in PR(latest commit):
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DDM([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [3]: %timeit ddm.extract_slice(slice(9998, 9999, None), slice(9998, 9999, None))
181 ms ± 16.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Timings for the code suggested above by you and after modification https://github.com/sympy/sympy/pull/21118#discussion_r602465259
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DDM([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [3]: %timeit ddm.extract_slice(slice(9998, 9999, None), slice(9998, 9999, None))
5.14 µs ± 623 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
> Rather than guessing can you do some timings?

Sorry about that.

Timings for the code already in PR(latest commit):
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DDM([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [3]: %timeit ddm.extract_slice(slice(9998, 9999, None), slice(9998, 9999, None))
181 ms ± 16.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Timings for the code suggested above by you and after modification https://github.com/sympy/sympy/pull/21118#discussion_r602465259
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DDM([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [3]: %timeit ddm.extract_slice(slice(9998, 9999, None), slice(9998, 9999, None))
5.14 µs ± 623 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
Okay, that's better. You should also time for different combinations of large and small matrices and slices e.g. a large slice from a large matrix and a small slice from a small matrix or slicing a single column or row etc.

Reply:
The `extract_slice` method should only be for slicing and not for getting a single element.

It looks like the implementation above would turn a 1x1 matrix slice into a scalar which is not correct.

Reply:
Also there should not be imports inside a performance sensitive function. If there are problems with cyclic imports then the import can go at the bottom of the module. There shouldn't be though if `DomainScalar` is defined in its own file.

Reply:
Actually `DDM` should not return `DomainScalar`. It should be `DomainMatrix` that wraps the result in `DomainScalar` if needed.

Reply:
> combinations of large and small matrices and slices e.g. a large slice from a large matrix and a small slice from a small matrix or slicing a single column or row etc.

Timings with a variety of DDMs:

Suggested code:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DDM([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [3]: %timeit ddm.extract_slice(slice(5, 9, None), slice(8, 9, None))
6.94 µs ± 585 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit ddm.extract_slice(slice(0, 9, None), slice(8, 9, None))
6.05 µs ± 495 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: ddm = DDM([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [6]: %timeit ddm.extract_slice(slice(9998, 9999, None), slice(9998, 9999, None))
4.56 µs ± 219 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit ddm.extract_slice(slice(0, 9999, None), slice(9998, 9999, None))
1.91 ms ± 276 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Code in the PR:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DDM([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [3]: %timeit ddm.extract_slice(slice(5, 9, None), slice(8, 9, None))
16.4 µs ± 316 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit ddm.extract_slice(slice(0, 9, None), slice(8, 9, None))
25.2 µs ± 6.42 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]: ddm = DDM([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [6]: %timeit ddm.extract_slice(slice(9998, 9999, None), slice(9998, 9999, None))
280 ms ± 40.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit ddm.extract_slice(slice(0, 9999, None), slice(9998, 9999, None))
### It kept running for 10 mins still no output. 
```
Clearly, the approach suggested by you is way faster than what is implemented as of now. Should it be implemented?

> The `extract_slice` method should only be for slicing and not for getting a single element.
> 
> It looks like the implementation above would turn a 1x1 matrix slice into a scalar which is not correct.

Sure, will add a DDM.getitem method the same as SDM for specifically returning single element.

Reply:
> combinations of large and small matrices and slices e.g. a large slice from a large matrix and a small slice from a small matrix or slicing a single column or row etc.

Timings with a variety of DDMs:

Suggested code:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DDM([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [3]: %timeit ddm.extract_slice(slice(5, 9, None), slice(8, 9, None))
6.94 µs ± 585 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit ddm.extract_slice(slice(0, 9, None), slice(8, 9, None))
6.05 µs ± 495 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: ddm = DDM([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [6]: %timeit ddm.extract_slice(slice(9998, 9999, None), slice(9998, 9999, None))
4.56 µs ± 219 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit ddm.extract_slice(slice(0, 9999, None), slice(9998, 9999, None))
1.91 ms ± 276 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Code in the PR:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DDM([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [3]: %timeit ddm.extract_slice(slice(5, 9, None), slice(8, 9, None))
16.4 µs ± 316 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit ddm.extract_slice(slice(0, 9, None), slice(8, 9, None))
25.2 µs ± 6.42 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]: ddm = DDM([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [6]: %timeit ddm.extract_slice(slice(9998, 9999, None), slice(9998, 9999, None))
280 ms ± 40.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit ddm.extract_slice(slice(0, 9999, None), slice(9998, 9999, None))
### It kept running for more than15 mins still no output. 
```
Clearly, the approach suggested by you is way faster than what is implemented as of now. Should it be implemented?

> The `extract_slice` method should only be for slicing and not for getting a single element.
> 
Sure will remove it. 
Should we introduce `DDM.getitem `as well, which can return a single item as scalar as `SDM.getitem `would do?


> Also there should not be imports inside a performance sensitive function. If there are problems with cyclic imports then the import can go at the bottom of the module. There shouldn't be though if `DomainScalar` is defined in its own file.

For now, in this PR `DomainScalar` is defined in `domainmatrix.py`, will fix it once [DomainScalar](https://github.com/sympy/sympy/pull/21120) gets merged.

Reply:
> Actually `DDM` should not return `DomainScalar`. It should be `DomainMatrix` that wraps the result in `DomainScalar` if needed.

I am not sure if I follow. 
Maybe I am missing something but why will an internal method of `DDM` would involve `DomainMatrix` as its return type?

@oscarbenjamin, Kindly share your opinion.
Thanks.

Reply:
Yes, there should be `DDM.getitem` so that `DDM` and `SDM` provide the same interface.

> Maybe I am missing something but why will an internal method of `DDM` would involve `DomainMatrix` as its return type?

The `DomainMatrix.__getitem__` method will call `DDM.getitem` which should give a domain element. The `DomainMatrix.__getitem__` method can then wrap the domain element in a `DomainScalar`.

Reply:
> Timings with a variety of DDMs:

Okay good. We should implement the faster approach.

Reply:
> The `DomainMatrix.__getitem__` method will call `DDM.getitem` which should give a domain element. The `DomainMatrix.__getitem__` method can then wrap the domain element in a `DomainScalar`.

For now, we have implemented `DDM.extract_slice` independent of `DomainMatrix.__getitem__`. Should this be made how `SDM.extract_slice` works, i.e. it assumes `DomainMatrix.__getitem__` will call it and not the user will call it explicitly.

Kindly share your opinion.
Thanks.